### PR TITLE
Added flags for tagging periodic boundary faces.

### PR DESCRIPTION
**Pull request by tarikdzanic**

### CODE REVIEW

File: pyfr/readers/base.py [hunk 1]
```diff
@@ -138,6 +138,13 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
         pfaces = defaultdict(list)
         nodepts = self._nodepts
 
+        # Setup flags for periodic faces 
+        pfaceflags = {}
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: Why do we need two loops here?

File: pyfr/readers/base.py [hunk 2]
```diff
@@ -150,13 +157,21 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
...
                 for lfn, rfn in zip(lfnodes[lfidx], rfnodes[rfidx]):
-                    lf = resid.pop(tuple(sorted(lfn)))
-                    rf = resid.pop(tuple(sorted(rfn)))
+                    lf = list(resid.pop(tuple(sorted(lfn))))
+                    rf = list(resid.pop(tuple(sorted(rfn))))
 
-                    pfaces[pftype].append([lf, rf])
+                    # Add periodic face flags
+                    if lpent in pfaceflags:
+                        lf[-1] = pfaceflags[lpent]
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: When is this condition false?

File: pyfr/readers/base.py (hunk 1 above)
Comment (tarikdzanic) [reply pyfr/readers/base.py]: Updated

File: pyfr/readers/base.py [hunk 3]
```diff
@@ -182,8 +188,8 @@ def get_connectivity(self):
         # Pair the fluid-fluid faces
         fpairs, resid = self._pair_fluid_faces(ffaces)
 
-        # Identify periodic boundary face pairs
-        pfpairs = self._pair_periodic_fluid_faces(bpart, resid)
+        # Tag and pair periodic boundary faces
+        pfpairs = self._pair_periodic_fluid_faces(bpart, resid)
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: Extra whitespace.

File: pyfr/readers/base.py [hunk 4]
```diff
@@ -150,13 +150,19 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
                 rfidx = fuzzysort(rfpts.mean(axis=1).T, range(len(rfnodes)))
 
                 for lfn, rfn in zip(lfnodes[lfidx], rfnodes[rfidx]):
-                    lf = resid.pop(tuple(sorted(lfn)))
-                    rf = resid.pop(tuple(sorted(rfn)))
+                    lf = list(resid.pop(tuple(sorted(lfn))))
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: `resid.pop(tuple(sorted(lfn)))[:-1] + (int(k) + 1,)` allows you to work with tuples directly.

File: pyfr/readers/base.py [hunk 5]
```diff
@@ -138,7 +138,7 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
         pfaces = defaultdict(list)
         nodepts = self._nodepts
 
-        for lpent, rpent in self._pfacespents.values():
+        for k, [lpent, rpent] in self._pfacespents.items():
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: Customary to use a tuple here.

**Comment by FreddieWitherden** [comment]
One thing we also may want to check is if the pents regular exceed 127.

**Comment by tarikdzanic** [comment]
> One thing we also may want to check is if the pents regular exceed 127.

Do you want to raise a runtime error or just print a warning and resume without flagging the boundaries exceeding 127?

**Comment by FreddieWitherden** [comment]
> > One thing we also may want to check is if the pents regular exceed 127.
> 
> Do you want to raise a runtime error or just print a warning and resume without flagging the boundaries exceeding 127?

Another option would be to simply persist these as 16- or 32-bit integers.  It would only require changes in 3/4 places.

**Comment by tarikdzanic** [comment]
Would that affect backwards compatibility with meshes imported beforehand? Because the i1 type seems to pop up in the partitioner and the base system.

**Comment by FreddieWitherden** [comment]
> Would that affect backwards compatibility with meshes imported beforehand? Because the i1 type seems to pop up in the partitioner and the base system.

So we'd replace all of the i1's with i2's or i4's.  Everything would be upward compatible since it is valid to cast an i1 to an i2 or i4 and so old meshes would continue to function.

File: pyfr/readers/base.py [hunk 6]
```diff
@@ -150,13 +150,18 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
                 rfidx = fuzzysort(rfpts.mean(axis=1).T, range(len(rfnodes)))
 
                 for lfn, rfn in zip(lfnodes[lfidx], rfnodes[rfidx]):
-                    lf = resid.pop(tuple(sorted(lfn)))
-                    rf = resid.pop(tuple(sorted(rfn)))
+                    # Add periodic face flags
+                    # Left = +, right = -, periodic BC number offset plus 1 to avoid +/- 0
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: Move the left/right bit down to just above lf and rf (but keep the +1 bit).

File: pyfr/readers/base.py [hunk 7]
```diff
@@ -150,13 +150,18 @@ def _pair_periodic_fluid_faces(self, bpart, resid):
...
+                    flg = int(k) + 1
+
+                    lf = resid.pop(tuple(sorted(lfn)))[:-1] + ( flg,)
+                    rf = resid.pop(tuple(sorted(rfn)))[:-1] + (-flg,)
 
                     pfaces[pftype].append([lf, rf])
 
         return pfaces
-
+
```
Comment (FreddieWitherden) [inline pyfr/readers/base.py]: ?

**Comment by FreddieWitherden** [comment]
@GiorgioGiangaspero would be good to get your feedback here.

**Comment by FreddieWitherden** [comment]
It's also worth checking if the pent can ever be negative.

**Comment by tarikdzanic** [comment]
> It's also worth checking if the pent can ever be negative.

Current implementation in the Gmsh and CGNS readers does not allow it:

`p = re.match(r'periodic[ _-]([a-z0-9]+)[ _-](l|r)$', name)`

**Pull request merged**